Darwin: Prohibit static initializers in Matter.framework

ksperling-apple opened the pull request:
Globals should either be "constinit" (i.e. use a constrexpr constructor) and trivially destructible, or use the `Global<>` / `AtomicGlobal<>` helpers.

ksperling-apple committed:
Darwin: Prohibit static initializers in Matter.framework

Globals should either be "constinit" (i.e. use a constrexpr constructor) and
trivially destructible, or use the Global<> / AtomicGlobal<> helpers.

bzbarsky-apple commented on src/messaging/ReliableMessageProtocolConfig.cpp:
```diff
@@ -62,14 +67,15 @@ namespace {
 // This is not a static member of ReliableMessageProtocolConfig because the free
 // function GetLocalMRPConfig() needs access to it.
-Optional<ReliableMessageProtocolConfig> sDynamicLocalMPRConfig;
+// Use std::optional to avoid a static initializers
```
```suggestion
// Use std::optional to avoid a static initializer
```

ksperling-apple committed:
Update src/messaging/ReliableMessageProtocolConfig.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

ksperling-apple committed:
Use correct value in ReliableMessageProtocolConfig unit test override

ksperling-apple committed:
Enable -no_inits for release builds only

ASAN and TSAN both use initializers, so enabling it for Debug builds breaks
those in CI. Ideally we could just turn it off for builds that actually use
*SAN but that probably requires migrating the project to use xcconfig files.
